### ISSUE TITLE
onFileAdded: cannot read property 'mtime' of undefined

### ISSUE THREAD

**Issue by ahfarmer** [webpack/watchpack#9]
I get this error occasionally in `onFileAdded` when using webpack `compiler.watch`. I haven't been able to track down an exact repro. 

```
project/node_modules/webpack/node_modules/watchpack/lib/DirectoryWatcher.js:211
        this.setFileTime(filePath, +stat.mtime, false, "add");
                                        ^
TypeError: Cannot read property 'mtime' of undefined
    at DirectoryWatcher.onFileAdded (project/node_modules/webpack/node_modules/watchpack/lib/DirectoryWatcher.js:211:34)
    at EventEmitter.emit (events.js:98:17)
    at EventEmitter.<anonymous> (project/node_modules/webpack/node_modules/watchpack/node_modules/chokidar/index.js:135:15)
    at project/node_modules/webpack/node_modules/watchpack/node_modules/chokidar/index.js:145:7
    at Object.oncomplete (fs.js:107:15)
```

I wonder if it might be related to this commit?
https://github.com/webpack/watchpack/commit/9681996a1ce5d58036f188557055fb7ac109246d

I see that chokidar is supposed to always provide the `stat` value, so I don't see how this could happen. Maybe it is a problem with chokidar?

**Comment by sokra** [comment]
@es128

**Comment by es128** [comment]
Using the `alwaysStat` option should ensure the results of `fs.stat` are always passed with events. Maybe `fs.stat` failed to provide a stat object? Unfortunately without a repro case I don't think I have any other insight about this.

**Comment by es128** [comment]
Chokidar doesn't validate the results from `fs.stat`, just passes it along. You may be experiencing a race condition here with the file disappearing between operations or something like that. Probably best to just handle the case where `stat` is `undefined`

https://github.com/paulmillr/chokidar/blob/master/index.js#L143-L146

**Comment by ahfarmer** [comment]
@sokra would you accept a pull request in which I check if stat is undefined before accessing stat.mtime? If stat is undefined, I'll pass null as the 2nd argument to setFileTime.

**Comment by sokra** [comment]
@es128 chokidar doesn't check the `error` parameter [here](https://github.com/paulmillr/chokidar/blob/master/index.js#L143-L146). Is this intended? Should be expected that chokidar gives use no stats object? Or should chokidar supress the event instead (or convert it to a `removed` event)?

**Comment by es128** [comment]
@sokra this particular stat call is just sort of tacked on to support `alwaysStat` - it does not currently impact any other behavior. Since chokidar does not eagerly `stat` files by default, this feature was added as a convenience so implementations wouldn't all need similar logic checking whether `stat` was passed and doing their own `fs.stat` if not. 

Your suggestion may be feasible, but it has the potential to mean completely different events depending upon whether that option is enabled. I'm not sure whether chokidar can make a proper decision on behalf of all implementations what should happen when this condition is encountered, but we could consider it. 

This is the first report of this problem with `alwaysStat`.

**Comment by ahfarmer** [comment]
If it helps, I do have a consistent reproduction of this issue now: https://gist.github.com/ahfarmer/fd0ddeb1692dc11cda80

Just download the gist, `npm install`, then `npm run start`. You'll see it printing `Stats is undefined.` repeatedly, and then you can look in `index.js` to see that `stats` is indeed undefined. **Mac only**

@es128 maybe we could add a note to the docs saying that `alwaysStat` is really an 'almostAlwaysStat' and you can't count on `stats` always being defined, but when `stats` is undefined, the file in question has likely been removed.

@sokra would you be OK with a pull request in which we ignore events for which 'stat' is undefined? An undefined `stats` indicates that the added file no longer exists.

I'm happy to put together a pull request (in either or both projects) for whichever approach you guys think is best.

**Comment by es128** [comment]
Actually I'm thinking it does make sense to change chokidar's behavior here to simply suppress the event in this situation.

I'm assuming there's no need to convert it to an `unlink` event as that will still separately be reported by the system and responded to by chokidar.

@ahfarmer can you try a simple patch at https://github.com/paulmillr/chokidar/blob/master/index.js#L143-L146 that does something like `if (error || !stats) return;` and let us know what impact you observe?

**Comment by ahfarmer** [comment]
Nice! I think suppressing the event sounds like a good call to me.

Just tried it. My issue is resolved when I add your code verbatim as line 145:

    fs.stat(path, function(error, stats) {
      if (error || !stats) return;
      ...

What's the next step? Do you need anything else from me?

**Comment by es128** [comment]
Up to you if you want to PR it and get your name in the chokidar commit history :)

Otherwise I'll take care of it shortly.

**Comment by ahfarmer** [comment]
Oooo, ok, I've submitted a PR: https://github.com/paulmillr/chokidar/pull/299

**Comment by es128** [comment]
Landed and published in chokidar@1.0.3. 

@sokra you should be able to close this issue.
